### PR TITLE
Add integer range condition and integer `start_from` to interface

### PR DESCRIPTION
**Pull request by coszio**
To finish taking advantage of the newly added range interface. This adds:
- `Range<IntPayloadType>` to `RangeInterface`
- `StartFrom::Int(IntPayloadType)` variant for order-by `scroll` requests.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/types.rs [hunk 1]
```diff
@@ -59,6 +59,34 @@ impl DateTimeWrapper {
     }
 }
 
+impl FromStr for DateTimePayloadType {
```
Comment (coszio) [inline lib/segment/src/types.rs:62]: This part was just moved closer to the definition of `DateTimeWrapper`

File: lib/segment/src/payload_storage/condition_checker.rs [hunk 2]
```diff
@@ -140,18 +137,34 @@ impl ValueChecker for Match {
...
+
+impl ValueChecker for Range<IntPayloadType> {
+    fn check_match(&self, payload: &Value) -> bool {
+        payload
+            .as_number()
+            .and_then(|x| x.as_i64().or_else(|| x.as_f64().map(|x| x as i64)))
+            .map_or(false, |x| self.check_range(x))
+    }
+}
+
```
Comment (coszio) [inline lib/segment/src/payload_storage/condition_checker.rs:167]: For unindexed fields, we now try to convert into the type of the range to do the checks

File: lib/segment/src/index/query_optimization/condition_converter.rs [hunk 3]
```diff
@@ -245,6 +246,28 @@ pub fn get_float_range_checkers(
...
+            Some(Box::new(move |point_id: PointOffsetType| {
+                num_index
+                    .get_values(point_id)
+                    .is_some_and(|values| values.iter().copied().any(|f| range.check_range(f)))
+            }))
+        }
+        _ => None,
+    }
+}
+
```
Comment (coszio) [inline lib/segment/src/index/query_optimization/condition_converter.rs:270]: For indexed fields, we follow same approach as before, converting the range into the type of the index

**Comment by coszio** [comment]
I am facing issues making appropriate conversions in the python client (in which pydantic is matching floats to ints, and viceversa). Maybe this is a good reason to **not** include these changes. 

In the meantime, I'll draft this PR to avoid merging just yet.

File: lib/segment/src/index/field_index/numeric_index/mod.rs [hunk 4]
```diff
@@ -526,21 +522,27 @@ impl ValueIndexer<FloatPayloadType> for NumericIndex<FloatPayloadType> {
...
-        range: &RangeInterface,
+        range_interface: &RangeInterface,
     ) -> Box<dyn DoubleEndedIterator<Item = (T, PointOffsetType)> + '_> {
-        let range = match range {
-            RangeInterface::Float(float_range) => float_range.map(T::from_f64),
-            RangeInterface::DateTime(datetime_range) => {
-                datetime_range.map(|dt| T::from_i64(dt.timestamp()))
-            }
-        };
+        let range = Range::from(range_interface.clone());
```
Comment (xzfc) [inline lib/segment/src/index/field_index/numeric_index/mod.rs:545]: Nit: maybe we could avoid cloning if there is `impl<T: Numericable> From<&RangeInterface> for Range<T>` (note the `&`).

File: lib/segment/src/payload_storage/condition_checker.rs [hunk 5]
```diff
@@ -140,18 +137,34 @@ impl ValueChecker for Match {
...
+            .and_then(|x| x.as_f64().or_else(|| x.as_i64().map(|x| x as f64)))
+            .map_or(false, |x| self.check_range(x))
+    }
+}
+
+impl ValueChecker for Range<IntPayloadType> {
+    fn check_match(&self, payload: &Value) -> bool {
+        payload
+            .as_number()
+            .and_then(|x| x.as_i64().or_else(|| x.as_f64().map(|x| x as i64)))
```
Comment (xzfc) [inline lib/segment/src/payload_storage/condition_checker.rs:163]: This rounds floats down, so, e.g., `{"payload": 0.3}` would be rounded to `0i64` and then matched by `{lte: Some(0i64)}` which is not what we want.
I think a similar rounding issue could also be in the case above (`x as f64`).
I propose to do comparison using `num-cmp`, like discussed in https://github.com/qdrant/qdrant/pull/3533#pullrequestreview-1864712419. I.e. explicitly handle cases `i64<i64`, `i64<f64`, `f64<i64`, and so on.

**Review by xzfc — CHANGES_REQUESTED** [review]

File: lib/segment/src/payload_storage/condition_checker.rs (hunk 5 above)
Comment (coszio) [reply lib/segment/src/payload_storage/condition_checker.rs:163]: You're totally right, I'll update accordingly

Comment (coszio) [reply lib/segment/src/payload_storage/condition_checker.rs]: @xzfc I have added a new check_range implementation for the cases where we need to check different types.

For the case of `Range<f64>`, though, I've restored its original implementation, as all integers were getting converted to floats via `as f64` inside the `serde_json` crate already.

File: lib/segment/src/index/field_index/numeric_index/mod.rs (hunk 4 above)
Comment (coszio) [reply lib/segment/src/index/field_index/numeric_index/mod.rs:545]: Yeah, I also thought about that, but I think this is a fairly inexpensive operation.

To make this more explicit, I have just pushed a change to make `Range` and `RangeInterface` `Copy`

**Review by xzfc — DISMISSED** [review]

**Comment by generall** [comment]
Please feel free to make another iteration, once we will have a better idea of how integrate it into the client

**Pull request closed without merging**